doc/specs/#476 - Package Pinning.md

```diff
@@ -0,0 +1,205 @@
+---
+author: Yao Sun @yao-msft
+created on: 2022-10-12
+last updated: 2022-10-12
+issue id: 476
+---
+
+# Package Pinning
+
+For [#476](https://github.com/microsoft/winget-cli/issues/476)
+
+## Abstract
+
+This spec describes the functionality and high level implementation design of Package Pinning feature.
+
+## Inspiration
+
+This is inspired by functionalities in other package managers, as well as community feedback.
+- Packages may introduce breaking changes that users may not want integrate into their workflow quite yet.
+- Packages may update themselves so that it will be duplicate effort for winget to try to update them.
+- User may want to maintain some of the packages through other channels outside of winget, or prefer one source over others within winget.
+- User may want some of the packages to stay in some major versions but allow minor version changes during upgrade.
+
+## Solution Design
+
+#### Package Pinning types
+To achieve goals listed above, winget will support 3 types of Package Pinning:
+- **Blocking:** The package is blocked from `winget upgrade --all` or `winget upgrade <specific package>`, user has to unblock the package to let winget perform upgrade.
+- **Pinning:** The package is excluded from `winget upgrade --all` but allowed in `winget upgrade <specific package>`, a new argument `--include-pinned` will be introduced to let `winget upgrade --all` to include pinned packages.
+- **Gating:** The package is pinned to specific version(s). For example, if a package is pinned to version `1.2.*`, any version between `1.2.0` to `1.2.<anything>` is considered valid.
```

Comment:
We need to define how each type of pin interacts with `RequireExplicitUpgrade`. I'm guessing something like this:
* Pinning pins and `RequireExplicitUpgrade` behave the same way. Even though they do the same things, we allow pinning in case the manifest gets updated to remove the field or the package updates to a version without that field.
* Blocking pins override `RequireExplicitUpgrade`, so we won't update even if the include flag is added. We have to both remove the blocking pin and pass the flag when updating.
* Gating pins block from updating beyond the gated version, even with the special flag for `RequireExplicitUpgrade`, and updates within that version range still require the special flag.

Does that sound good?